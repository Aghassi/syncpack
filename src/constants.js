export const DEFAULT_PACKAGES = './packages/*/package.json';
export const DEFAULT_SOURCE = './package.json';
