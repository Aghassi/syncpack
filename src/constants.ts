import { IManifestKey } from './typings';

export const FIX_MISMATCHES = {
  description: 'set dependencies used with different versions to the same version',
  name: 'fix-mismatches'
};
export const LIST = {
  description: 'list every dependency used in your packages',
  name: 'list'
};
export const LIST_MISMATCHES = {
  description: 'list every dependency used with different versions in your packages',
  name: 'list-mismatches'
};
export const DEFAULT_PATTERN = './packages/*/package.json';
export const DEPENDENCY_TYPES: IManifestKey[] = ['dependencies', 'devDependencies', 'peerDependencies'];
export const GREATER = 1;
export const LESSER = -1;
export const OPTION_PACKAGES = {
  description: `location of packages, defaults to "${DEFAULT_PATTERN}"`,
  spec: '-p, --packages <glob>'
};
export const SAME = 0;
export const SEMVER_ORDER = ['<', '<=', '', '~', '^', '>=', '>', '*'];
