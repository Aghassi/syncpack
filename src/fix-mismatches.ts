import chalk from 'chalk';
import { writeJson } from 'fs-extra';
import _ = require('lodash');
import { relative } from 'path';
import {
  OPTION_INDENT,
  OPTION_SOURCES,
  OPTIONS_DEV,
  OPTIONS_PEER,
  OPTIONS_PROD
} from './constants';
import { collect } from './lib/collect';
import { getDependencyTypes } from './lib/get-dependency-types';
import { getIndent } from './lib/get-indent';
import { getPackages } from './lib/get-packages';
import { getMismatchedVersionsByName } from './lib/get-versions-by-name';
import { getNewest } from './lib/version';
import { CommanderApi } from './typings';

export const run = async (program: CommanderApi) => {
  program
    .option(OPTION_SOURCES.spec, OPTION_SOURCES.description, collect)
    .option(OPTIONS_PROD.spec, OPTIONS_PROD.description)
    .option(OPTIONS_DEV.spec, OPTIONS_DEV.description)
    .option(OPTIONS_PEER.spec, OPTIONS_PEER.description)
    .option(OPTION_INDENT.spec, OPTION_INDENT.description)
    .parse(process.argv);

  const pkgs = getPackages(program);
  const dependencyTypes = getDependencyTypes(program);
  const indent = getIndent(program);
  const mismatchedVersionsByName = getMismatchedVersionsByName(
    dependencyTypes,
    pkgs
  );

  Object.entries(mismatchedVersionsByName).forEach(([name, versions]) => {
    const newest = getNewest(versions);
    pkgs.forEach(({ data, path }) => {
      dependencyTypes.forEach((type) => {
        if (data[type] && data[type][name] && data[type][name] !== newest) {
          console.log(
            relative(process.cwd(), path),
            name,
            data[type][name],
            '->',
            newest
          );
          data[type][name] = newest;
        }
      });
    });
  });

  await Promise.all(
    pkgs.map(({ data, path }) => writeJson(path, data, { spaces: indent }))
  );

  _.each(pkgs, (pkg) => {
    console.log(chalk.blue(`./${relative('.', pkg.path)}`));
  });
};
