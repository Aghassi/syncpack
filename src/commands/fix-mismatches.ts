import chalk from 'chalk';
import { writeFileSync } from 'fs-extra';
import { EOL } from 'os';
import { relative } from 'path';
import { DependencyType, SyncpackConfig } from '../constants';
import { getDependencyTypes } from './lib/get-dependency-types';
import { getHighestVersion } from './lib/get-highest-version';
import { getMismatchedDependencies } from './lib/get-installations';
import { getWrappers, SourceWrapper } from './lib/get-wrappers';
import { log } from './lib/log';

type Options = Pick<SyncpackConfig, 'dev' | 'filter' | 'indent' | 'peer' | 'prod' | 'source'>;

export const fixMismatches = (dependencyTypes: DependencyType[], filter: RegExp, wrappers: SourceWrapper[]): void => {
  const iterator = getMismatchedDependencies(dependencyTypes, wrappers);
  const mismatches = Array.from(iterator).filter(({ name }) => name.search(filter) !== -1);

  mismatches.forEach((installedPackage) => {
    const versions = installedPackage.installations.map((installation) => installation.version);
    const newest = getHighestVersion(versions);
    if (newest !== null) {
      installedPackage.installations.forEach(({ type, name, source }) => {
        const dependencies = source.contents[type];
        if (dependencies) {
          dependencies[name] = newest;
        }
      });
    }
  });
};

export const fixMismatchesToDisk = ({ dev, filter, indent, peer, prod, source }: Options): void => {
  const toJson = (wrapper: SourceWrapper): string => `${JSON.stringify(wrapper.contents, null, indent)}${EOL}`;
  const dependencyTypes = getDependencyTypes({ dev, peer, prod });
  const wrappers = getWrappers({ source });
  const allBefore = wrappers.map(toJson);

  fixMismatches(dependencyTypes, filter, wrappers);

  wrappers.forEach((wrapper, i) => {
    const shortPath = relative(process.cwd(), wrapper.filePath);
    const before = allBefore[i];
    const after = toJson(wrapper);
    if (before !== after) {
      writeFileSync(wrapper.filePath, after);
      log(chalk.green('âœ“'), shortPath);
    } else {
      log(chalk.dim('-', shortPath));
    }
  });
};
