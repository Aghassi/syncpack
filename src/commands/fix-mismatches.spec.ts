import * as mock from '../../test/mock';
import { DEFAULT_CONFIG } from '../constants';
import * as api from './fix-mismatches';

describe('fixMismatches', () => {
  let fixMismatches: typeof api.fixMismatches;

  afterEach(() => {
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    jest.mock('./lib/log', () => ({ log: jest.fn() }));
    fixMismatches = require('./fix-mismatches').fixMismatches;
  });

  describe('when dependencies are installed with different versions', () => {
    describe('when the dependency is not a package maintained in this workspace', () => {
      it('uses the highest version', () => {
        const wrappers = [mock.wrapper('a', ['foo@0.1.0']), mock.wrapper('b', [], ['foo@0.2.0'])];
        fixMismatches(wrappers, DEFAULT_CONFIG);
        expect(wrappers).toMatchSnapshot();
      });
    });

    describe('when the dependency is a package maintained in this workspace', () => {
      it('uses the workspace version', () => {
        const wrappers = [
          mock.wrapper('a', ['foo@0.1.0']),
          mock.wrapper('b', [], ['foo@0.2.0']),
          mock.wrapper('foo', [], [], [], { name: 'foo', version: '0.0.1' }),
        ];
        fixMismatches(wrappers, DEFAULT_CONFIG);
        expect(wrappers).toMatchSnapshot();
      });
    });
  });

  it('ignores non-semver dependencies', () => {
    const wrappers = [
      mock.wrapper('a', ['foo@link:vendor/foo-0.1.0']),
      mock.wrapper('b', ['foo@link:vendor/foo-0.2.0']),
    ];
    fixMismatches(wrappers, DEFAULT_CONFIG);
    expect(wrappers).toMatchSnapshot();
  });
});
