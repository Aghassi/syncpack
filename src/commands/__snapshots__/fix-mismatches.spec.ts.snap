// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixMismatches ignores non-semver dependencies 1`] = `
Array [
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "link:vendor/foo-0.1.0",
      },
    },
    "filePath": "/a/package.json",
  },
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "link:vendor/foo-0.2.0",
      },
    },
    "filePath": "/b/package.json",
  },
]
`;

exports[`fixMismatches when dependencies are installed with different versions when the dependency is a package maintained in this workspace uses the workspace version 1`] = `
Array [
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.0.1",
      },
    },
    "filePath": "/a/package.json",
  },
  Object {
    "contents": Object {
      "devDependencies": Object {
        "foo": "0.0.1",
      },
    },
    "filePath": "/b/package.json",
  },
  Object {
    "contents": Object {
      "name": "foo",
      "version": "0.0.1",
    },
    "filePath": "/foo/package.json",
  },
]
`;

exports[`fixMismatches when dependencies are installed with different versions when the dependency is not a package maintained in this workspace uses the highest version 1`] = `
Array [
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.2.0",
      },
    },
    "filePath": "/a/package.json",
  },
  Object {
    "contents": Object {
      "devDependencies": Object {
        "foo": "0.2.0",
      },
    },
    "filePath": "/b/package.json",
  },
]
`;
